src/main.rs

```diff
@@ -1,52 +1,151 @@
+        if input.update(&event) {
+            // Close events
+            if input.key_pressed(VirtualKeyCode::Escape) || input.quit() {
+                *control_flow = ControlFlow::Exit;
+                return;
+            }
+
+            if input.mouse_pressed(0) {
+                trace!("Pressed LMB");
+                world.is_left_mouse_button_held_down = true
+            } else if input.mouse_released(0) {
+                trace!("Released LMB");
+                world.is_left_mouse_button_held_down = false
+            }
+
+            if input.mouse_pressed(1) {
+                trace!("Pressed RMB");
+                world.is_right_mouse_button_held_down = true
+            } else if input.mouse_released(1) {
+                trace!("Released RMB");
+                world.is_right_mouse_button_held_down = false
+            }
+
+            if let Some(xy) = input.mouse() {
+                world.mouse_xy = xy;
+            }
+
+            // Resize the window
+            if let Some(size) = input.window_resized() {
+                pixels.resize_surface(size.width, size.height);
+            }
+
+            // Update internal state and request a redraw
+            window.request_redraw();
+            world.update(&pixels, frame_time);
 
-const MODEL_HEIGHT: usize = 300;
-const MODEL_WIDTH: usize = (MODEL_HEIGHT as f32 * ASPECT_RATIO) as usize;
+            frame_time = time_of_last_frame_start.elapsed().as_secs_f32();
+            time_of_last_frame_start = Instant::now();
 
-const HEIGHT_RATIO: f32 = WINDOW_HEIGHT as f32 / MODEL_HEIGHT as f32;
-const WIDTH_RATIO: f32 = WINDOW_WIDTH as f32 / MODEL_WIDTH as f32;
+            frame_counter += 1;
 
-const CURRENT_MODEL: (f32, f32) = model_presets::BRAIN_CORAL;
+            if time_of_last_fps_counter_update.elapsed().as_secs() > 1 {
+                time_of_last_fps_counter_update = Instant::now();
+                let _ = fps_values.push(frame_counter);
+                frame_counter = 0;
 
-struct MainState {
-    frames: usize,
-    reaction_diffusion_system: ReactionDiffusionSystem,
-    fast_forward: bool,
-    gradient: ColorGradient,
-    is_mouse_button_pressed: bool,
+                let fps_sum: i32 = fps_values.iter().sum();
+                let avg_fps = fps_sum as f32 / fps_values.len() as f32;
+                info!("FPS {}", avg_fps.trunc());
+            }
+        }
+    });
+}
+
+struct World {
+    pub fast_forward: bool,
+    pub frames: usize,
+    pub gradient: ColorGradient,
+    pub is_left_mouse_button_held_down: bool,
+    pub is_right_mouse_button_held_down: bool,
+    pub previous_frame_start_time: Instant,
+    pub reaction_diffusion_system: ReactionDiffusionSystem,
+    pub mouse_xy: (f32, f32),
+    pub previous_mouse_xy: Option<(f32, f32)>,
 }
 
-impl MainState {
-    fn new(_ctx: &mut Context) -> GameResult<MainState> {
-        let s = MainState {
+impl World {
+    fn new() -> Self {
+        Self {
+            fast_forward: false,
```

Comment:
This is still here but it looks like the implementation of fast forward got deleted in this PR.